### PR TITLE
add image decoding network with upsampling+conv

### PR DESCRIPTION
**Pull request by hnyu**
Added another implementation of image decoding network which uses upsampling and conv layers. This implementation can avoid checkerboard artifacts as commonly seen in decoders with transposed conv layers. https://distill.pub/2016/deconv-checkerboard

In short, for transposed conv layers, depending on the kernel size and stride, some locations in the next layer might have field overlaps from several locations in the previous layer. This overlap results in a greater default magnitude. Sometimes it's difficult for the network learns to remove this bias.

### CODE REVIEW

File: alf/networks/encoding_networks.py [hunk 1]
```diff
@@ -232,6 +232,129 @@ def __init__(self,
...
+            start_decoding_channels: the initial number of channels we'd
+                like to have for the feature map. Note that we always first
+                project an input latent vector into a vector of an appropriate
+                length so that it can be reshaped into (``start_decoding_channels``,
+                ``start_decoding_height``, ``start_decoding_width``).
+            upsampling_mode: the argument for choosing an upsampling algorithm
+                for ``torch.nn.Upsample``.
+            same_padding: similar to TF's conv2d ``same`` padding mode. If
+                True, the user provided paddings in ``transconv_layer_params`` will
+                be replaced by automatically calculated ones; if False, it
```
Comment (emailweixu) [inline alf/networks/encoding_networks.py:297]: It's better to simply describe how the padding is calculated for same_padding=True and False

Comment (hnyu) [reply alf/networks/encoding_networks.py:297]: > It's better to simply describe how the padding is calculated for same_padding=True and False

``ImageEncodingNetwork`` already has a paragraph explaining the two padding modes. I've added a sentence referring to its docstring.

**Review by emailweixu — APPROVED** [review]
